pkg/targetgroupbinding/networking_manager.go

```diff
@@ -2,36 +2,348 @@ package targetgroupbinding
-	return nil
+// fetchTGBsWithNetworking returns all targetGroupsBindings with networking rules in cluster.
+func (m *defaultNetworkingManager) fetchTGBsWithNetworking(ctx context.Context) (map[types.NamespacedName]*elbv2api.TargetGroupBinding, error) {
+	tgbList := &elbv2api.TargetGroupBindingList{}
+	if err := m.k8sClient.List(ctx, tgbList); err != nil {
+		return nil, err
+	}
+	tgbWithNetworkingByKey := make(map[types.NamespacedName]*elbv2api.TargetGroupBinding, len(tgbList.Items))
+	for i := range tgbList.Items {
+		tgb := &tgbList.Items[i]
+		if tgb.Spec.Networking != nil {
+			tgbWithNetworkingByKey[k8s.NamespacedName(tgb)] = tgb
+		}
+	}
+	return tgbWithNetworkingByKey, nil
 }
 
-func (m *defaultResourceManager) computeSInstanceSGsForNodePortEndpoints(ctx context.Context, endpoints []backend.NodePortEndpoint) []string {
-	return nil
+// computeDesiredIngressIPPermissionsBySG will compute the desired Ingress IPPermissions per SecurityGroup.
+// It will also GC unnecessary entries in endpointSGsByTGB, and return whether all tgb have endpointSGs specified.
+func (m *defaultNetworkingManager) computeDesiredIngressIPPermissionsBySG(tgbWithNetworkingByKey map[types.NamespacedName]*elbv2api.TargetGroupBinding) (map[string][]networking.IPPermissionInfo, bool, error) {
+	tgbNetworkingsBySG := make(map[string][]elbv2api.TargetGroupBindingNetworking)
+	for tgbKey, endpointSGs := range m.endpointSGsByTGB {
+		tgb, exists := tgbWithNetworkingByKey[tgbKey]
+		if !exists {
+			delete(m.endpointSGsByTGB, tgbKey)
+			continue
+		}
+		for _, endpointSG := range endpointSGs {
+			tgbNetworkingsBySG[endpointSG] = append(tgbNetworkingsBySG[endpointSG], *tgb.Spec.Networking)
+		}
+	}
+	computedEndpointSGsForAllTGBs := len(tgbWithNetworkingByKey) == len(m.endpointSGsByTGB)
+	ipPermissionsBySG := make(map[string][]networking.IPPermissionInfo)
+	for sgID, tgbNetworkings := range tgbNetworkingsBySG {
+		ipPermissions, err := m.computeDesiredIngressIPPermissions(tgbNetworkings)
+		if err != nil {
+			return nil, false, err
+		}
+		ipPermissionsBySG[sgID] = ipPermissions
+	}
+	return ipPermissionsBySG, computedEndpointSGsForAllTGBs, nil
+}
+
+func (m *defaultNetworkingManager) computeDesiredIngressIPPermissions(tgbNetworkings []elbv2api.TargetGroupBindingNetworking) ([]networking.IPPermissionInfo, error) {
+	var ipPermissions []networking.IPPermissionInfo
+	opts := cmp.Options{
+		ec2equality.CompareOptionForIPPermission(),
+		cmpopts.IgnoreFields(networking.IPPermissionInfo{}, "Labels"),
+	}
+
+	for _, tgbNetworking := range tgbNetworkings {
+		for _, rule := range tgbNetworking.Ingress {
+			for _, port := range rule.Ports {
+				for _, peer := range rule.From {
+					ipPermission, err := m.computeDesiredIngressIPPermission(port, peer)
+					if err != nil {
+						return nil, err
+					}
+					containsPermission := false
+					for _, permission := range ipPermissions {
```

Comment:
what is the expected number of permissions in the worst case? 
This could cause a potential bottleneck.

Reply:
humm, one securityGroup only have at most 10X permissions.
calculations like these will never be bottleneck :D (networking is is)